### PR TITLE
[Dynatrace v2] Ensure synchronization when taking snapshots

### PR DESCRIPTION
**Pull request by pirgeo**
We found a timing bug that can appear when using the Dynatrace v2 exporter, usually only when recording data at high volumes. In previous versions, recording a snapshot was not a synchronized operation. That means that it was possible to write data to a DynatraceSummary, while the snapshot was being created. In very rare cases this can lead to invalid data. This PR moves the synchronization while taking a snapshot to the innermost level, and ensures that no writes can happen while a Snapshot is produced.

### CODE REVIEW

**Comment by pirgeo** [comment]
@shakuzen or @jonatan-ivanov, could you take a look? I would love to get this into the next patch release, if possible 😄

**Review by jonatan-ivanov — APPROVED** [review]

File: implementations/micrometer-registry-dynatrace/src/main/java/io/micrometer/dynatrace/types/DynatraceTimer.java [hunk 1]
```diff
@@ -67,7 +71,7 @@ public DynatraceSummarySnapshot takeSummarySnapshot() {
 
     @Override
     public DynatraceSummarySnapshot takeSummarySnapshot(TimeUnit unit) {
-        return new DynatraceSummarySnapshot(min(unit), max(unit), totalTime(unit), count());
+        return convertIfNecessary(unit, summary.takeSummarySnapshot());
```
Comment (jonatan-ivanov) [inline implementations/micrometer-registry-dynatrace/src/main/java/io/micrometer/dynatrace/types/DynatraceTimer.java:74]:
```suggestion
        return convertIfNecessary(summary.takeSummarySnapshot(), unit);
```

File: implementations/micrometer-registry-dynatrace/src/main/java/io/micrometer/dynatrace/types/DynatraceTimer.java [hunk 2]
```diff
@@ -77,9 +81,17 @@ public DynatraceSummarySnapshot takeSummarySnapshotAndReset() {
 
     @Override
     public DynatraceSummarySnapshot takeSummarySnapshotAndReset(TimeUnit unit) {
-        DynatraceSummarySnapshot snapshot = takeSummarySnapshot(unit);
-        summary.reset();
-        return snapshot;
+        return convertIfNecessary(unit, summary.takeSummarySnapshotAndReset());
```
Comment (jonatan-ivanov) [inline implementations/micrometer-registry-dynatrace/src/main/java/io/micrometer/dynatrace/types/DynatraceTimer.java:84]:
```suggestion
        return convertIfNecessary(summary.takeSummarySnapshotAndReset(), unit);
```

File: implementations/micrometer-registry-dynatrace/src/test/java/io/micrometer/dynatrace/types/DynatraceTimerTest.java [hunk 3]
```diff
@@ -159,6 +135,48 @@ void testDifferentTimeUnits() {
...
+    @Test
+    void testConvertIfNecessary() {
+        TimeUnit unit = TimeUnit.MILLISECONDS;
+        DynatraceTimer timer = new DynatraceTimer(ID, CLOCK, DISTRIBUTION_STATISTIC_CONFIG, PAUSE_DETECTOR, unit);
+
+        DynatraceSummarySnapshot snapshot = new DynatraceSummarySnapshot(1000., 2000., 3000., 2);
+        DynatraceSummarySnapshot unconverted = timer.convertIfNecessary(unit, snapshot);
+        assertThat(unconverted).isEqualTo(snapshot);
+
+        DynatraceSummarySnapshot converted = timer.convertIfNecessary(TimeUnit.SECONDS, snapshot);
```
Comment (jonatan-ivanov) [inline implementations/micrometer-registry-dynatrace/src/test/java/io/micrometer/dynatrace/types/DynatraceTimerTest.java:147]:
```suggestion
        DynatraceSummarySnapshot unconverted = timer.convertIfNecessary(snapshot, unit);
        assertThat(unconverted).isEqualTo(snapshot);

        DynatraceSummarySnapshot converted = timer.convertIfNecessary(snapshot, TimeUnit.SECONDS);
```

**Review by jonatan-ivanov — COMMENTED** [review]
Since I applied the parameter change suggestion, the build is broken and I can't fix it on the branch but I think I can fix it with applying these suggestions. :)

**Review by jonatan-ivanov — APPROVED** [review]

**Comment by jonatan-ivanov** [comment]
@pirgeo Thank you for fixing this!
Could you please do two more things?

1. Could you please add a note in the description that with this change recordings can be blocked during publishing for a very small amount of time while we get and reset the values for the snapshot?
2. Could you please add some example error message that Dynatrace will produce if this happens so users who get it and search on it might discover this issue?

**Comment by pirgeo** [comment]
@jonatan-ivanov done! Should I send another PR for adding the missing test that you pointed out?

**Pull request merged**